tokio-executor/src/task/core.rs

```diff
@@ -0,0 +1,153 @@
+    /// Hot task state data
+    pub(super) header: Header<S>,
+
+    /// Either the future or output, depending on the execution stage.
+    pub(super) core: Core<T>,
+
+    /// Cold data
+    pub(super) trailer: Trailer,
+}
+
+/// The core of the task.
+///
+/// Holds the future or output, depending on the stage of execution.
+pub(super) struct Core<T: Future> {
+    stage: Stage<T>,
+}
+
+/// Crate public as this is also needed by the pool.
+#[repr(C)]
+pub(crate) struct Header<S: 'static> {
+    /// Task state
+    pub(super) state: State,
+
+    /// Pointer to the executor owned by the task
+    pub(super) executor: CausalCell<Option<NonNull<S>>>,
+
+    /// Pointer to next task, used for misc task linked lists.
+    pub(crate) queue_next: UnsafeCell<*const Header<S>>,
+
+    /// Pointer to the next task in the ownership list.
+    pub(crate) owned_next: UnsafeCell<Option<NonNull<Header<S>>>>,
+
+    /// Pointer to the previous task in the ownership list.
+    pub(crate) owned_prev: UnsafeCell<Option<NonNull<Header<S>>>>,
+
+    /// Table of function pointers for executing actions on the task.
+    pub(super) vtable: &'static Vtable<S>,
+
+    /// Used by loom to track the causality of the future. Without loom, this is
+    /// unit.
+    pub(super) future_causality: CausalCell<()>,
+}
+
+/// Cold data is stored after the future.
+pub(super) struct Trailer {
+    /// Consumer task waiting on completion of this task.
+    pub(super) waker: CausalCell<MaybeUninit<Option<Waker>>>,
+}
+
+/// Either the future or the output.
+enum Stage<T: Future> {
+    Running(Track<T>),
+    Finished(Track<super::Result<T::Output>>),
+    Consumed,
+}
+
+impl<T: Future, S: Schedule> Cell<T, S> {
+    /// Allocate a new task cell, containing the header, trailer, and core
+    /// structures.
+    pub(super) fn new(future: T, state: State) -> Box<Cell<T, S>> {
+        Box::new(Cell {
+            header: Header {
+                state,
+                executor: CausalCell::new(None),
+                queue_next: UnsafeCell::new(ptr::null()),
+                owned_next: UnsafeCell::new(None),
+                owned_prev: UnsafeCell::new(None),
+                vtable: raw::vtable::<T, S>(),
+                future_causality: CausalCell::new(()),
+            },
+            core: Core {
+                stage: Stage::Running(Track::new(future)),
+            },
+            trailer: Trailer {
+                waker: CausalCell::new(MaybeUninit::new(None)),
+            },
+        })
+    }
+}
+
+impl<T: Future> Core<T> {
+    pub(super) fn transition_to_consumed(&mut self) {
+        self.stage = Stage::Consumed
+    }
+
+    pub(super) fn poll<S>(&mut self, header: &Header<S>) -> Poll<T::Output>
+    where
+        S: Schedule,
+    {
+        let res = {
+            let future = match &mut self.stage {
+                Stage::Running(tracked) => tracked.get_mut(),
+                _ => panic!("unexpected stage"),
```

Comment:
Nit: should this be
```suggestion
                _ => unreachable!("unexpected stage"),
```
(and maybe print which stage we were in?)